### DISCUSSION TITLE
Disabling external header code analysis when using integrate install with VS2022?

### DISCUSSION THREAD

**Discussion by micka190** [microsoft/vcpkg#22157] [Q&A]
I'm using Visual Studio 2022 with vcpkg. I've run the `vcpkg integrate install` command to use libraries inside VS2022 **without using CMake** (nothing against CMake, but some of the packages I'm using have less than stellar support for it, and I'd rather they "just work" for these side projects).

I'm running into an issue with code analysis where it gives me hundreds of warnings from external library headers.
For example, the `libnoise` package has 269 warnings.

I've followed the steps in this [C++ Team Blog post](https://devblogs.microsoft.com/cppblog/customized-warning-levels-and-code-analysis-for-external-headers/) to disable code analysis in external headers, but it seems that vcpkg headers aren't recognized as external headers, because it isn't working. I've also tried adding them manually to my "External Include Directory" options under my project's Properties, but that hasn't done it either.

I'm guessing vcpkg appends its include directories at the end of the regular include directories list?

Is there any way to disable code analysis for external headers added by vcpkg when using `integrate install`? Or do I have to use CMake and configure it manually through that?

___

Here are some screenshots.

Here's what my External Include Directories look like:

<img width="572" alt="External Includes (1 or 2)" src="https://user-images.githubusercontent.com/63615275/147100274-cb1e72f5-bfa2-4a4f-a64d-5f3bf7ad8cce.PNG">
<img width="257" alt="External Includes (2 or 2)" src="https://user-images.githubusercontent.com/63615275/147100299-e3bfd80a-5707-4434-bb9e-a6ca1b342273.PNG">

Here are my External Includes settings:

<img width="316" alt="Disabled Code Analysis" src="https://user-images.githubusercontent.com/63615275/147100091-8a0f7233-0bdd-494f-b367-f607ce8d852e.PNG">

But I still get warnings:

<img width="502" alt="Warnings" src="https://user-images.githubusercontent.com/63615275/147100422-91d3359c-b75d-473a-b331-9658917cd2dc.PNG">

**Comment by DownerCase** [comment]
Raising this one from the dead a bit but I've run up against the same problem in a different situation. I tried turning on [`enableMicrosoftCodeAnalysis`](https://learn.microsoft.com/en-us/cpp/build/cmake-presets-json-reference?view=msvc-170) in my CMakePreset Vendor map, set a ruleset and enabled `disableExternalAnalysis` to prevent excessive warnings from code that isn't mine. I was very disappointed to then get flooded with warnings when running the analysis...

Interestingly Ninja correctly marks the vcpkg include as external and disables warnings.
`INCLUDES = <snip> -external:IF:\dev\<snip>\vcpkg_installed\x64-windows\include -external:W0`

However, when running the code analysis it executes:
```
cl " "/DWIN32" "/D_WINDOWS" \
"/GR" "/EHsc" "/Ob0" "/Od" "/RTC1" "-MDd" \
"/IF:\dev<snip>\vcpkg_installed\x64-windows\include" \
"F:\dev\<snip>\src\test.cpp" \
"/analyze:projectdirectoryF:\dev\<snip>" "/external:W0" "/external:env:CAExternalPath" "/analyze:external-" \
"/analyze:rulesetdirectoryC:\Program Files\Microsoft Visual Studio\2022\CommunityRelease\Team Tools\Static Analysis Tools\Rule Sets" \
"/analyze:rulesetC:\Program Files\Microsoft Visual Studio\2022\CommunityRelease\Team Tools\Static Analysis Tools\Rule Sets\NativeRecommendedRules.ruleset" \
"/c" "/Zs" "/errorReport:none" "/analyze" "/analyze:logF:\dev\<snip>\CodeAnalysis\results.sarif" "/DCODE_ANALYSIS" \
"/analyze:pluginC:\Program Files\Microsoft Visual Studio\2022\CommunityRelease\VC\Tools\MSVC\14.36.32532\bin\HostX64\x64\EspXEngine.dll"
```

My CMake Preset vendor map:
```
"microsoft.com/VisualStudioSettings/CMake/1.0": {
          "enableMicrosoftCodeAnalysis": true,
          "codeAnalysisRuleset": "NativeRecommendedRules.ruleset",
          "disableExternalAnalysis": true
}
```

Using Visual Studio 17.6.5 and latest vcpkg

**Reply by crud89** [reply]
Just experienced the same issue using latest vcpkg and VS 17.11.5. It appears to only affect certain headers, so maybe this is related to individual ports, but I am not sure. Did you find any solution to this?

**Reply by DownerCase** [reply]
No, I gave up and relied on clang-tidy instead. 

Wasn't even sure which tool was at fault vcpkg, CMake, or VS.
Would be nice to have it work though. A minimal repro with the latest tools would, as always, be a good start for sure.

**Reply by RT2Code** [reply]
I have the same issue, I reported it here: https://developercommunity.visualstudio.com/t/External-headers-are-not-excluded-for-co/10898359?q=cmake+external+header

**Reply by RT2Code** [reply]
They've acknowledged the issue and classified it as a bug. I'll report back here once it's fixed.

**Reply by DownerCase** [reply]
Awesome! Thanks for reporting it. Fingers crossed 🤞

**Reply by michailf** [reply]
This seems to work in the latest VS 2022
```
  <ItemDefinitionGroup>
    <ClCompile>
      <DisableAnalyzeExternal>true</DisableAnalyzeExternal>
      <ExternalWarningLevel>TurnOffAllWarnings</ExternalWarningLevel>
      <AdditionalOptions>/external:I "$(VcpkgInstalledDir.TrimEnd('\'))" %(AdditionalOptions)</AdditionalOptions>
    </ClCompile>
  </ItemDefinitionGroup>
```

This is my vcpkg.props that I include into my projects.  I think the key is the `ExternalWarningLevel` setting.

**Reply by RT2Code** [reply]
It may work for MSBuild projects, but the issue is still not fixed for CMake projects.
